File: Swift/ExampleTests/common/FrameValuesTyped.swift
public struct FrameValuesTyped: Equatable, CustomStringConvertible {
    public let frame: Int
    public let roll1: String
    public let roll2: String
    public let roll3: String
    public let score: String
    public let totalScore: String

    public init(frame: Int, roll1: String, roll2: String, roll3: String, score: String, totalScore: String) {
        self.frame = frame
        self.roll1 = roll1
        self.roll2 = roll2
        self.roll3 = roll3
        self.score = score
        self.totalScore = totalScore
    }

    public init(from s: FrameValuesString) {
        self.frame = Int(s.frame) ?? 0
        self.roll1 = s.roll1
        self.roll2 = s.roll2
        self.roll3 = s.roll3
        self.score = s.score
        self.totalScore = s.totalScore
    }

    public func toStringStruct() -> FrameValuesString {
        return FrameValuesString(
            frame: String(describing: frame),
            roll1: String(describing: roll1),
            roll2: String(describing: roll2),
            roll3: String(describing: roll3),
            score: String(describing: score),
            totalScore: String(describing: totalScore)
        )
    }

    public static func toStringList(_ list: [FrameValuesTyped]) -> [FrameValuesString] {
        return list.map { $0.toStringStruct() }
    }

    public static func fromStringList(_ list: [FrameValuesString]) -> [FrameValuesTyped] {
        return list.map { FrameValuesTyped(from: $0) }
    }

    public func toJSONValue() -> [String: Any] {
        return [
            "frame": frame,
            "roll1": roll1,
            "roll2": roll2,
            "roll3": roll3,
            "score": score,
            "totalScore": totalScore,
        ]
    }

    public func toJSON() throws -> String {
        return try Json.write(toJSONValue())
    }

    public init(fromJSONValue m: [String: Any]) throws {
        self.frame = try Json.asInt(Json.require(m, "frame"), "frame")
        self.roll1 = try Json.asString(Json.require(m, "roll1"), "roll1")
        self.roll2 = try Json.asString(Json.require(m, "roll2"), "roll2")
        self.roll3 = try Json.asString(Json.require(m, "roll3"), "roll3")
        self.score = try Json.asString(Json.require(m, "score"), "score")
        self.totalScore = try Json.asString(Json.require(m, "totalScore"), "totalScore")
    }

    public init(fromJSON text: String) throws {
        try self.init(fromJSONValue: Json.parseObject(text))
    }

    public static func toJSONList(_ list: [FrameValuesTyped]) throws -> String {
        return try Json.write(list.map { $0.toJSONValue() })
    }

    public static func fromJSONList(_ text: String) throws -> [FrameValuesTyped] {
        return try Json.parseArray(text).map {
            try FrameValuesTyped(fromJSONValue: Json.asObject($0, "FrameValuesTyped"))
        }
    }

    public var description: String {
        return "Frame=\(frame), Roll1=\(roll1), Roll2=\(roll2), Roll3=\(roll3), Score=\(score), TotalScore=\(totalScore)"
    }
}

File: Rust/ExampleTests/common/framevalues_typed.rs
#![allow(dead_code, unused_imports, unused_variables)]

use super::json;
use super::framevalues_string::FrameValuesString;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct FrameValuesTyped {
    pub frame: i32,
    pub roll1: String,
    pub roll2: String,
    pub roll3: String,
    pub score: String,
    pub totalscore: String,
}

impl FrameValuesTyped {
    pub fn from_str_struct(s: &FrameValuesString) -> Self {
        Self {
            frame: s.frame.parse::<i32>().unwrap_or_default(),
            roll1: s.roll1.clone(),
            roll2: s.roll2.clone(),
            roll3: s.roll3.clone(),
            score: s.score.clone(),
            totalscore: s.totalscore.clone(),
        }
    }

    pub fn to_str_struct(&self) -> FrameValuesString {
        FrameValuesString {
            frame: self.frame.to_string(),
            roll1: self.roll1.clone(),
            roll2: self.roll2.clone(),
            roll3: self.roll3.clone(),
            score: self.score.clone(),
            totalscore: self.totalscore.clone(),
        }
    }

    pub fn to_string_list(list: &[FrameValuesTyped]) -> Vec<FrameValuesString> {
        list.iter().map(|t| t.to_str_struct()).collect()
    }

    pub fn from_string_list(list: &[FrameValuesString]) -> Vec<FrameValuesTyped> {
        list.iter().map(FrameValuesTyped::from_str_struct).collect()
    }

    pub fn to_json_value(&self) -> json::Value {
        json::Value::Object(vec![
            ("Frame".to_string(), json::Value::number_from_i64(self.frame as i64)),
            ("Roll1".to_string(), json::Value::Str(self.roll1.clone())),
            ("Roll2".to_string(), json::Value::Str(self.roll2.clone())),
            ("Roll3".to_string(), json::Value::Str(self.roll3.clone())),
            ("Score".to_string(), json::Value::Str(self.score.clone())),
            ("TotalScore".to_string(), json::Value::Str(self.totalscore.clone())),
        ])
    }

    pub fn to_json(&self) -> String {
        json::write(&self.to_json_value())
    }

    pub fn from_json_value(v: &json::Value) -> json::JsonResult<Self> {
        Ok(Self {
            frame: json::as_i32(json::require(v, "Frame")?, "Frame")?,
            roll1: json::as_string(json::require(v, "Roll1")?, "Roll1")?,
            roll2: json::as_string(json::require(v, "Roll2")?, "Roll2")?,
            roll3: json::as_string(json::require(v, "Roll3")?, "Roll3")?,
            score: json::as_string(json::require(v, "Score")?, "Score")?,
            totalscore: json::as_string(json::require(v, "TotalScore")?, "TotalScore")?,
        })
    }

    pub fn from_json(text: &str) -> json::JsonResult<Self> {
        Self::from_json_value(&json::parse(text)?)
    }

    pub fn to_json_list(list: &[FrameValuesTyped]) -> String {
        json::write(&json::Value::Array(
            list.iter().map(|item| item.to_json_value()).collect(),
        ))
    }

    pub fn from_json_list(text: &str) -> json::JsonResult<Vec<FrameValuesTyped>> {
        let root = json::parse(text)?;
        let items = json::as_array(&root, "FrameValuesTyped")?;
        let mut result = Vec::with_capacity(items.len());
        for e in items {
            result.push(Self::from_json_value(e)?);
        }
        Ok(result)
    }
}
